### PR TITLE
Render Certificates in Program content language

### PR DESCRIPTION
**Pull request by McKenzieW**
Language is explicitly configured for a particular program's certificates
Template rendering language is set to the configured content_language
[LEARNER-1936](https://openedx.atlassian.net/browse/LEARNER-1936)

### CODE REVIEW

**Comment by McKenzieW** [comment]
@edx/learner-heart take a look

File: credentials/settings/base.py [hunk 1]
```diff
@@ -172,6 +173,8 @@
         }
     },
 ]
+# I'm not entirely sure whether this should live here or in the configuration repo
```
Comment (jlajoie) [inline credentials/settings/base.py]: I took a look at some of the models in edx-platform, it looks like they just include the settings in the model itself, might be best to just do that?

https://github.com/edx/edx-platform/blob/41862cf163bec5b6ac9cc2d7432ce887740e09f9/common/djangoapps/student/admin.py#L40 for example

File: credentials/templates/base.html [hunk 2]
```diff
@@ -5,7 +5,8 @@
 
 <!DOCTYPE html>
 <html>
-<head lang="{{ language_code }}">
+{% language content_language %}
+<head lang="{% if content_language %}{{content_language}}{% else %}{{language_code}}{% endif %}">
```
Comment (jlajoie) [inline credentials/templates/base.html]: I think you could shorten this to `content_language or language_code`

Comment (McKenzieW) [reply credentials/templates/base.html]: Would that look like this? 
lang="{{content_language}}{% or %}{{language_code}}"

File: credentials/apps/credentials/views.py [hunk 3]
```diff
@@ -87,6 +87,7 @@ def get_context_data(self, **kwargs):
             'user_credential': user_credential,
             'user_data': user_data,
             'child_templates': self.get_child_templates(),
+            'content_language': user_credential.credential.language,
```
Comment (Mjloturco) [inline credentials/apps/credentials/views.py]: Can this return null if no language is defined?If so , content_language of None might behave weirdly at https://github.com/edx/credentials/pull/311/files#diff-0ed0d978b3addaff36a2abbdc8fa12cfR8?

File: credentials/settings/base.py (hunk 1 above)
Comment (McKenzieW) [reply credentials/settings/base.py]: Actually what I was wondering was whether the dictionary should be here, in the credentials repo settings folder, or if it should be here, https://github.com/edx/configuration/blob/master/playbooks/roles/credentials/defaults/main.yml

File: credentials/apps/credentials/views.py (hunk 3 above)
Comment (McKenzieW) [reply credentials/apps/credentials/views.py]: If no language is defined then the language tag just doesn't do anything

File: credentials/templates/base.html (hunk 2 above)
Comment (Mjloturco) [reply credentials/templates/base.html]: The firstof tag might help here https://github.com/edx/credentials/pull/311/files#diff-0ed0d978b3addaff36a2abbdc8fa12cfR8

File: credentials/settings/base.py [hunk 4]
```diff
@@ -172,6 +173,8 @@
...
+CERTIFICATE_LANGUAGES = {'en': 'English', 'es_419': 'Español'}
```
Comment (bderusha) [inline credentials/settings/base.py]: We should default to just english or nothing in the base settings file

File: credentials/settings/base.py (hunk 1 above)
Comment (bderusha) [reply credentials/settings/base.py]: We should add this to the configuration repo (both the generic one as well as our own internal configuration)

File: credentials/templates/base.html [hunk 5]
```diff
@@ -5,7 +5,8 @@
 
 <!DOCTYPE html>
 <html>
-<head lang="{{ language_code }}">
+{% language content_language %}
+<head lang="{% firstof content_language language_code %}">
```
Comment (amangano-edx) [inline credentials/templates/base.html]: Why does `<head>` have a lang attribute but `<html>` doesn't? I think `<html>` is likely the better place for it. If children of <html> use a different language, they can always define their own lang attribute.

File: credentials/templates/base.html [hunk 6]
```diff
@@ -5,7 +5,8 @@
 
 <!DOCTYPE html>
 <html>
-<head lang="{{ language_code }}">
+{% language content_language %}
```
Comment (amangano-edx) [inline credentials/templates/base.html]: This probably belongs above `<html>`.

File: credentials/templates/base.html [hunk 7]
```diff
@@ -37,5 +38,6 @@
 {% include '_analytics.html' %}
 
 {% block javascript %}{% endblock %}
+{% endlanguage %}
```
Comment (amangano-edx) [inline credentials/templates/base.html]: This probably belongs below `</html>`

File: credentials/templates/base.html (hunk 5 above)
Comment (amangano-edx) [reply credentials/templates/base.html]: Also, where does `language_code` come from?

Comment (amangano-edx) [reply credentials/templates/base.html]: Will this still result in us showing certificates in the browser-prefered language if we haven't explicitly set a certificate language in the configuration? Is that desirable behavior?

File: credentials/apps/credentials/forms.py [hunk 8]
```diff
@@ -18,6 +20,14 @@ class Meta:
 
 
 class ProgramCertificateAdminForm(forms.ModelForm):
+    def __init__(self, *args, **kwargs):
+        super(ProgramCertificateAdminForm, self).__init__(*args, **kwargs)
+        languages = settings.CERTIFICATE_LANGUAGES.items()
+        lang_choices = sorted(languages, key=itemgetter(1))
```
Comment (amangano-edx) [inline credentials/apps/credentials/forms.py]: What does `itemgetter` do?

File: credentials/apps/credentials/forms.py [hunk 9]
```diff
@@ -18,6 +20,14 @@ class Meta:
 
 
 class ProgramCertificateAdminForm(forms.ModelForm):
+    def __init__(self, *args, **kwargs):
+        super(ProgramCertificateAdminForm, self).__init__(*args, **kwargs)
+        languages = settings.CERTIFICATE_LANGUAGES.items()
```
Comment (amangano-edx) [inline credentials/apps/credentials/forms.py]: Nit: It might be good to just define this as a list of tuples in the configuration rather than as a dictionary if we plan to use it like that.

File: credentials/apps/credentials/views.py (hunk 3 above)
Comment (amangano-edx) [reply credentials/apps/credentials/views.py]: Is the behavior of the language tag when no language is provided defined as part of the interface, or does it just happen to work that way? If it's not part of the defined interface, it may be good to guard against passing `None` or `''` to `{% language %}` in case something about the implementation changes in the next version of Django. Maybe we could move the `or` logic from https://github.com/edx/credentials/pull/311/files#diff-0ed0d978b3addaff36a2abbdc8fa12cfR9 to here and then update the template to always expect content_language to be set?

**Comment by amangano-edx** [comment]
Is it possible to test the language rendering

File: credentials/apps/credentials/views.py [hunk 10]
```diff
@@ -87,6 +87,7 @@ def get_context_data(self, **kwargs):
             'user_credential': user_credential,
```
Comment (amangano-edx) [inline credentials/apps/credentials/views.py]: Would it be possible to add a test for this view to verify that content_language is provided in the template context and that the template renders correctlu?

**Comment by amangano-edx** [comment]
Sorry, I must have mis-clicked something and closed this accidentally.

File: credentials/templates/base.html (hunk 5 above)
Comment (McKenzieW) [reply credentials/templates/base.html]: `language_code` is the browser-preferred language, which I would imagine would be the desired render language if the certificate language hasn't been explicitly set.
I can make this an attribute of `<html>` though instead of `<head>`. I only added the logic here because it was the existing location of the `lang` attribute.

File: credentials/apps/credentials/forms.py (hunk 8 above)
Comment (McKenzieW) [reply credentials/apps/credentials/forms.py]: It gets the first element of the tuples so as to sort the list by the the language as opposed to the language code

File: credentials/templates/base.html (hunk 5 above)
Comment (amangano-edx) [reply credentials/templates/base.html]: I think for a11y compliance the lang attribute should be on the `<html>` element. 

You should check with Jasper about what we should do if certificate language is not set for a program. I vaguely remember talking about turning off browser-detection for consistency.

File: credentials/apps/credentials/forms.py (hunk 9 above)
Comment (McKenzieW) [reply credentials/apps/credentials/forms.py]: For the configuration yaml file it is simpler to represent the data as a dictionary than as a list of tuples

**Comment by McKenzieW** [comment]
@bderusha @amangano-edx what do you think?

File: credentials/apps/credentials/tests/test_views.py [hunk 11]
```diff
@@ -164,6 +165,24 @@ def test_logo_missing_exception(self):
         with self.assertRaisesMessage(MissingCertificateLogoError, 'No certificate image logo defined for program'):
             self._render_user_credential(use_proper_logo_url=False)
 
+    @ddt.data(
+        (True, 'Felicitaciones, '),
+        (False, 'Congratulations, ')
+    )
+    @ddt.unpack
+    @responses.activate
+    def test_render_language(self, langauge_set, expected_text):
```
Comment (amangano-edx) [inline credentials/apps/credentials/tests/test_views.py]: Will this pass if you do not have a Spanish translation for 'Congratulations'? It might be safer to use the built-in dummy language for testing, `eo`.

Comment (amangano-edx) [reply credentials/apps/credentials/tests/test_views.py]: Actually, it might be better to just check that the HTML lang element is set correctly.

File: credentials/templates/base.html [hunk 12]
```diff
@@ -4,8 +4,9 @@
 {% load render_bundle from webpack_loader %}
 
 <!DOCTYPE html>
-<html>
-<head lang="{{ language_code }}">
+{% language language %}
```
Comment (amangano-edx) [inline credentials/templates/base.html]: Are there any namespace issues here?

Comment (McKenzieW) [reply credentials/templates/base.html]: Because the variable name is the same as the method name? I didn't experience any issues with it in testing, but I can change it to be safe.

File: credentials/apps/credentials/models.py [hunk 13]
```diff
@@ -230,6 +230,11 @@ class ProgramCertificate(AbstractCertificate):
                   "feature will only be displayed in the certificate if the attribute 'Total hours of effort' has "
                   "been set for the program in Discovery."
     )
+    language = models.CharField(
+        max_length=8,
+        null=True,
+        help_text=u'Language in which certificates for this program will be rendered'
```
Comment (clintonb) [inline credentials/apps/credentials/models.py]: In Python 3.x, everything is Unicode. There is no need for the `u` prefix.
